flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FunctionDefinitionFactory.java

```diff
@@ -34,4 +34,29 @@ public interface FunctionDefinitionFactory {
      * @return a {@link FunctionDefinition}
      */
     FunctionDefinition createFunctionDefinition(String name, CatalogFunction catalogFunction);
```

Comment:
Add default implementation of the old method (users don't need to override it anymore).
And add a deprecation message on the old method to instruct users to implement the new method. 

https://github.com/apache/flink/pull/4616/files#diff-64d9c652ffc3c60b6d838200a24b106eeeda4b2d853deae94dbbdf16d8d694c2R51 is a good example for evolving methods.

Reply:
Should we add default implementation for the old method?  If old method has default implementaion, seems user must implement new method, which will bring Incompatibility problem.

Reply:
Good suggestion!